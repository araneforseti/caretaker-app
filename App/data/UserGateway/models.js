const Permissions = {
    FOCUS: 'focus',
    ADMIN: 'admin',
    HELPER: 'helper'
}

const NotificationTypes = {
    PUSH: 'push',
    EMAIL: 'email',
    SMS: 'sms'
}

const NotificationTriggers = {
    ANNOUNCEMENT: 'announcement',
    EVENT: 'event',
    EVENT_BY_ROLE: 'event by role'
}

const Settings = {
    [NotificationTypes.PUSH]: {
        name: 'Push',
        value: true
    },
    [NotificationTypes.EMAIL]: {
        name: 'Email',
        value: false
    },
    [NotificationTypes.SMS]: {
        name: 'SMS',
        value: false
    },
    [NotificationTriggers.ANNOUNCEMENT]: {
        name: 'Announcements',
        value: true
    },
    [NotificationTriggers.EVENT]: {
        name: 'Events',
        value: false
    },
    [NotificationTriggers.EVENT_BY_ROLE]: {
        name: 'Events specific to your role',
        value: true
    }
}

module.exports = {
  Permissions: Permissions,
  NotificationTypes: NotificationTypes,
  NotificationTriggers: NotificationTriggers,
  Settings: Settings
}
