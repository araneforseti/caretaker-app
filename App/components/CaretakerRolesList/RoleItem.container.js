import React from 'react';

export default container = (Component) => class extends React.Component {
  constructor() {
    super();
    this.state = {
      hasRole: null
    };
  }

  componentWillReceiveProps() {
    this.setState({
      hasRole: this.props.hasRole
    });
  }

  _onToggleSwitch = (value) => this.setState({ hasRole: value });

  _onLongPress = () => alert('Long Pressed');

  render() {
    const {
      name,
      description,
      isHelper,
      isLast
    } = this.props;

    return <Component
      name={name}
      description={description}
      isHelper={isHelper}
      isLast={isLast}
      onLongPress={this._onLongPress}
      onToggleSwitch={this._onToggleSwitch}
      hasRole={this.state.hasRole}
    />;
  }

}
