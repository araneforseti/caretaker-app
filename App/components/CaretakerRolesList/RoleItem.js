import React from 'react';
import PropTypes from 'prop-types';
import {
  Switch,
  TouchableOpacity,
} from 'react-native';
import renderif from '../../utils/renderif';
import theme from '../../config/theme';

import container from './RoleItem.container';
import ListItem from '../material-ui/ListItem';

export class RoleItem extends React.PureComponent {
  render() {
    const {
      name,
      description,
      hasRole,
      isHelper,
      onLongPress,
      onToggleSwitch,
      isLast
    } = this.props;

    return <TouchableOpacity onLongPress={onLongPress}>
      <ListItem
        divider={isLast}
        centerElement={ {
          primaryText: name,
          secondaryText: description
        } }
        rightElement={renderif(isHelper, <Switch
          value={hasRole}
          onValueChange={onToggleSwitch}
          onTintColor={theme.palette.lightColor}
          thumbTintColor={hasRole ? theme.palette.secondaryColor : null}
        />)}
      />
    </TouchableOpacity>;
  }
}

RoleItem.propTypes = {
  name: PropTypes.string,
  description: PropTypes.string,
  hasRole: PropTypes.bool,
  isHelper: PropTypes.bool,
  isLast: PropTypes.bool,

//From Container
  onLongPress: PropTypes.func,
  onToggleSwitch: PropTypes.func
};

export default container(RoleItem);
