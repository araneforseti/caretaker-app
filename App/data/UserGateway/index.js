import { Record } from 'immutable';
import UserModels from './models.js'
import { API_URL } from '../../config/ApiConfig';

export const Permissions = UserModels.Permissions
export const NotificationTypes = UserModels.NotificationTypes
export const NotificationTriggers = UserModels.NotificationTriggers

const createUserSettings = () => ({
  [NotificationTypes.PUSH]: {
    name: 'Push',
    value: true
  },
  [NotificationTypes.EMAIL]: {
    name: 'Email',
    value: false
  },
  [NotificationTypes.SMS]: {
    name: 'SMS',
    value: false
  },
  [NotificationTriggers.ANNOUNCEMENT]: {
    name: 'Announcements',
    value: true
  },
  [NotificationTriggers.EVENT]: {
    name: 'Events',
    value: false
  },
  [NotificationTriggers.EVENT_BY_ROLE]: {
    name: 'Events specific to your role',
    value: true
  }
})

export const User = Record({
  name: 'Defaul User',
  identifier: 'default@domain.com',
  permission: Permissions.HELPER,
  roles: [],
  phone: '123-456-7890',
  settings: createUserSettings()
});

const createUserGateway = function() {
  let allUsers = null;
  let currentUser = null;

  return {
    getCurrentUser() {
      return currentUser;
    },

    getAll() {
      if (allUsers === null) {
        //api call goes here
        allUsers = [
          new User({ name: 'S', identifier: 's', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone: '123-456-7890', settings: createUserSettings() }),
          new User({ name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend'], settings: createUserSettings() }),
          new User({ name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone: '123-456-7890', settings: createUserSettings() })
        ]
      }
      return allUsers;
    },

    getName(identifier) {
      const user = allUsers.find((user) => user.identifier === identifier)
      return user ? user.name : 'TBD';
    },

    getRoles() {
      return currentUser.roles
    },

    login(email) {
      //make call to google oauth and store the results
      let results = {
        email: email
      };
      let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
      if (typeof userFromBackend !== 'undefined') {
        currentUser = userFromBackend;
        return true;
      }
      return false;
    },

    logout() {
      // make call to oauth and/or invalidate current session
      this.currentUser = null;
    },

    isSelf(identifier) {
      return currentUser.identifier === identifier;
    },

    isAdmin() {
      return currentUser.permission === Permissions.ADMIN ||
        currentUser.permission === Permissions.FOCUS;
    },

    getCurrentPermissions() {
      return currentUser.permission;
    },

    inviteUser(email) {
      //api call to invite the user
      alert('invited ' + email + '!');
    },

    updateSettings(settings) {
      //api call to update the user settings
    }
  }
}

export default createUserGateway();
