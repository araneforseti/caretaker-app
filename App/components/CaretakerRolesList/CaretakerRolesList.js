import React from 'react';
import PropTypes from 'prop-types';
import {
  View,
  FlatList
} from 'react-native';

import RoleItem from './RoleItem';
import container from './CaretakerRolesList.container';
import styles from './CaretakerRolesList.styles';

class CaretakerRolesList extends React.Component {
  _renderItem = ({item, index}) => <RoleItem
    name={item.name}
    description={item.description}
    isHelper={this.props.isHelper}
    hasRole={this.props.userHasRole(item.name)}
    isLast={index === this.props.caretakerRoles.length - 1}
  />;

  render() {
    return <View style={styles.container}>
      <FlatList
        data={this.props.caretakerRoles}
        dataWhichCouldChange={this.props.caretakerRoles.map(role => role.volunteer)}
        renderItem={this._renderItem}
        keyExtractor={(item) => item.id}
      />
    </View>;
  }
}

CaretakerRolesList.propTypes = {
//From container
  caretakerRoles: PropTypes.array,
  isHelper: PropTypes.bool,
  userHasRole: PropTypes.func
};

export default container(CaretakerRolesList);
