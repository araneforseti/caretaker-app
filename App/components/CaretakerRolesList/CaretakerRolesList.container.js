import React from 'react';
import CaretakerRolesGateway from '../../data/CaretakerRoles/CaretakerRoles.gateway';
import UserGateway from '../../data/UserGateway';

export default container = (List) => class extends React.Component {
  constructor() {
    super();
    this.state = {
      caretakerRoles: [],
      userRoles: []
    };
  }

  componentDidMount() {
    const userRoles = UserGateway.getRoles();

    let newState = Object.assign({}, this.state);
    CaretakerRolesGateway.getAll().then(caretakerRoles => {
      newState.caretakerRoles = caretakerRoles;
      newState.userRoles = userRoles;
      this.setState(newState);
    });
  }

  userHasRole = (role) => {
    const words = this.state.userRoles.map(v => v.toLowerCase());
    return words.indexOf(role.toLowerCase()) >= 0;
  };

  render() {
    return <List
      isHelper={this.props.isHelper}
      caretakerRoles={this.state.caretakerRoles}
      userHasRole={this.userHasRole}
    />;
  }
}
