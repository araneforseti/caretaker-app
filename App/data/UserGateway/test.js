import UserGateway from './index.js';
import { Record } from 'immutable';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    });

    it('should get the current user as an immutable record', () => {
        UserGateway.login('s');

        const user =  UserGateway.getCurrentUser();

        //Record.isRecord is in the immutable@4.0.0 release candidate
        expect(user instanceof Record).toBe(true);
    })
});
