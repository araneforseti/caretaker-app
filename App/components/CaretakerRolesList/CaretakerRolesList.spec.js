import React from 'react';
import renderer from 'react-test-renderer';
import CaretakerRolesList from './CaretakerRolesList';
import UserGateway, { Permissions } from '../../data/UserGateway';

describe('<CaretakerRolesList />', () => {
  let props;
  const caretakerRolesList = () => renderer.create(<CaretakerRolesList {...props} />);

  beforeEach(() => {
    props = {
      isHelper: undefined
    };

    jest.spyOn(UserGateway, 'getRoles').mockReturnValue(['driver']);
  });

  it('should render correctly', () => {
    props.isHelper = false;
    expect(caretakerRolesList()).toMatchSnapshot();
  });

  describe('userHasRole', () => {
    it('should return true when the user has the given role', () => {
      expect(caretakerRolesList().getInstance().userHasRole('driver')).toBe(true);
    });

    it('should return false when the user does not have the given role', () => {
      expect(caretakerRolesList().getInstance().userHasRole('active friend')).toBe(false);
    });
  });

});
