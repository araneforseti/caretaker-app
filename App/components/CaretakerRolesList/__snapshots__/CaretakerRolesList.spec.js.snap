// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CaretakerRolesList /> should render correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "description": "Gives rides to things",
          "id": 1,
          "name": "Driver",
        },
        Object {
          "description": "Helps coordinate people sign ups",
          "id": 2,
          "name": "Coordinator",
        },
        Object {
          "description": "Picks up groceries",
          "id": 3,
          "name": "Groceries",
        },
        Object {
          "description": "Gets focus out and active (eg, walks) during vulnerable times",
          "id": 4,
          "name": "Active Friend",
        },
        Object {
          "description": "Cooks food cause yum",
          "id": 5,
          "name": "Chef",
        },
      ]
    }
    dataWhichCouldChange={
      Array [
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onViewableItemsChanged={undefined}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <ThemeProvider
            uiTheme={
              Object {
                "palette": Object {
                  "accentColor": "#FF7575",
                  "alternateTextColor": "white",
                  "lightColor": "#FFC0C0",
                  "primaryColor": "#FF7575",
                  "primaryTextColor": "black",
                  "secondaryColor": "#CC5D5D",
                },
                "spacing": Object {
                  "miniActionButtonSize": 42,
                },
                "toolbar": Object {
                  "container": Object {
                    "height": 50,
                  },
                },
                "typography": Object {
                  "fontSize": Object {
                    "display": 25,
                  },
                  "subheading": Object {
                    "fontSize": 16,
                    "fontWeight": "normal",
                    "lineHeight": 24,
                  },
                },
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
